client/src/app/process-details/process-details.component.ts

```diff
@@ -15,45 +15,176 @@ import {NgxNode} from './ngxNode';
 export class ProcessDetailsComponent implements OnInit {
 
   @Input()
-  selfUrl: string;
+  processUrl: string;
 
   @Input()
   schemaUrl: string;
 
+  processId: string;
+
   inputBiomaterials: MetadataDocument[];
   protocols: MetadataDocument[];
   derivedBiomaterials: MetadataDocument[];
   derivedFiles: MetadataDocument[];
 
+  protocolsToAdd: MetadataDocument[] = [];
+  inputBiomaterialsToAdd: MetadataDocument[] = [];
+  outputBiomaterialsToAdd: MetadataDocument[] = [];
+  outputFilesToAdd: MetadataDocument[] = [];
+
   links: Link[] = [];
-  nodes: NgxNode[] = [{
-    id: 'process',
-    label: 'process'
-  }];
+  nodes: NgxNode[] = [];
   done: boolean;
 
   zoomToFit$: Subject<boolean> = new Subject();
+  update$: Subject<any> = new Subject();
 
   constructor(private ingestService: IngestService) {
   }
 
   ngOnInit(): void {
+    this.processId = this.getIdFromUrl(this.processUrl);
+    this.refreshGraph();
+  }
+
+  updateGraph() {
+    this.update$.next(true);
+  }
+
+  getConcreteType(metadata: MetadataDocument): string {
+    return metadata.content['describedBy'].split('/').pop();
+  }
+
+  fitGraph() {
+    this.zoomToFit$.next(true);
+  }
+
+  onProtocolPicked($event: MetadataDocument) {
```

Comment:
All of these on* event handlers are the same more or less and can be refactored. Same goes for remove*, add*, and the ones in ingest.sevice (I just didn't get round to refactoring them yet).